### PR TITLE
Finished onboarding challenge

### PR DESCRIPTION
**Pull request by gracexuwt**
# Purpose
The purpose of this PR is to finish the Orbital Firmware onboarding: https://uworbital.notion.site/Firmware-Onboarding-Challenge-24340dfccb0547a5a7246ce6d7207c28#2d1d1b1b368c484fa725838805c14911

# New Changes
- Added functionality to driver functions, task, and OS handler

# Testing
- Ran the unit tests and integration test and passed.

# Outstanding Changes
- N/A

### CODE REVIEW

File: services/thermal_mgr/thermal_mgr.c [hunk 1]
```diff
@@ -42,19 +45,43 @@ void initThermalSystemManager(lm75bd_config_t *config) {
...
+  LOG_IF_ERROR_CODE(thermalMgrSendEvent(&event));
 }
 
 static void thermalMgr(void *pvParameters) {
-  /* Implement this task */
   while (1) {
-    
+    error_code_t errCode;
+    thermal_mgr_event_t event;
+    float temp;
```
Comment (Gjjjiang) [inline services/thermal_mgr/thermal_mgr.c]: usually, we want to minimize the scope of the variables. temp is only used inside the line 72's if statement scope.

File: services/thermal_mgr/thermal_mgr.c [hunk 2]
```diff
@@ -42,19 +45,43 @@ void initThermalSystemManager(lm75bd_config_t *config) {
...
+      LOG_IF_ERROR_CODE(readTempLM75BD(LM75BD_OBC_I2C_ADDR, &temp));
+
+      if(event.type == THERMAL_MGR_EVENT_MEASURE_TEMP_CMD) {
+        addTemperatureTelemetry(temp);
+      } else if(event.type == THERMAL_MGR_EVENT_INTERRUPT) {
+        if(temp > THERMAL_MGR_OVERTEMP) { //over temperature
+          overTemperatureDetected();
+        } else if(temp < THERMAL_MGR_HYSTERESIS) { //safe operating conditions
+          safeOperatingConditions();
+        }
```
Comment (Gjjjiang) [inline services/thermal_mgr/thermal_mgr.c]: Should make sure the temperature reading was successful before performing an action.

File: lm75bd/lm75bd.c [hunk 3]
```diff
@@ -26,8 +27,24 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
 }
 
 error_code_t readTempLM75BD(uint8_t devAddr, float *temp) {
-  /* Implement this driver function */
-  
+  error_code_t errCode;
+  float tempConst = 0.125f;
```
Comment (Gjjjiang) [inline lm75bd/lm75bd.c]: This can be a define constant.

**Review by Gjjjiang — CHANGES_REQUESTED** [review]
just some changes, but overall, good work.

File: lm75bd/lm75bd.c [hunk 4]
```diff
@@ -26,8 +29,23 @@ error_code_t lm75bdInit(lm75bd_config_t *config) {
...
+  uint8_t data[2] = {0};
+  RETURN_IF_ERROR_CODE(i2cReceiveFrom(devAddr, data, 2));
+
+  // Process register data
+  int16_t tempVal = (data[0] << 8) | data[1];
+  if((tempVal >> 15) && 0x1) { //Check D10, if 1 then negative
+    *temp = -((~tempVal >> 5) + 1) * TEMP_CONST; //invert two's complement
+  } else {
+    *temp = (tempVal >> 5) * TEMP_CONST;
+  }
```
Comment (Yarik-Popov) [inline lm75bd/lm75bd.c]: This logic is unnecessarily complex. Just bit shift 5he result then multiply by the temp const

File: services/thermal_mgr/thermal_mgr.c [hunk 5]
```diff
@@ -42,19 +45,47 @@ void initThermalSystemManager(lm75bd_config_t *config) {
 }
 
 error_code_t thermalMgrSendEvent(thermal_mgr_event_t *event) {
-  /* Send an event to the thermal manager queue */
+  if(event == NULL) return ERR_CODE_INVALID_ARG;
+  
+  error_code_t errCode;
+  if(xQueueSend(thermalMgrQueueHandle, event, 0) == errQUEUE_FULL) {
```
Comment (Yarik-Popov) [inline services/thermal_mgr/thermal_mgr.c:52]: Check that the queue isn't null, if so then return invalid state

File: services/thermal_mgr/thermal_mgr.c [hunk 6]
```diff
@@ -42,19 +45,47 @@ void initThermalSystemManager(lm75bd_config_t *config) {
...
 static void thermalMgr(void *pvParameters) {
-  /* Implement this task */
   while (1) {
-    
+    error_code_t errCode;
+    thermal_mgr_event_t event;
+
+    if(xQueueReceive(thermalMgrQueueHandle, &event, portMAX_DELAY) == pdTRUE) {
+      float temp;
+      errCode = readTempLM75BD(LM75BD_OBC_I2C_ADDR, &temp);
```
Comment (Yarik-Popov) [inline services/thermal_mgr/thermal_mgr.c]: In general, it is best to perform logic required by an event after checking the type. Lets say we add an event that doesn't require read the temperature, then reading the temperature when receiving this new event might cause issues.

File: services/thermal_mgr/thermal_mgr.c [hunk 7]
```diff
@@ -42,19 +45,47 @@ void initThermalSystemManager(lm75bd_config_t *config) {
...
+
+      if(errCode == ERR_CODE_SUCCESS) {
+        if(event.type == THERMAL_MGR_EVENT_MEASURE_TEMP_CMD) {
+          addTemperatureTelemetry(temp);
+        } else if(event.type == THERMAL_MGR_EVENT_INTERRUPT) {
```
Comment (Yarik-Popov) [inline services/thermal_mgr/thermal_mgr.c]: Additionally log an error if some other event is received

**Review by Yarik-Popov — CHANGES_REQUESTED** [review]
Overall good, not much to change

**Comment by Yarik-Popov** [comment]
Lgtm, dm me your email and I'll add you to the notion

**Pull request closed without merging**